common/scala/src/main/scala/org/apache/openwhisk/core/database/mongodb/MongoDBArtifactStore.scala

```diff
@@ -0,0 +1,661 @@
+      s"[ATT_GET] '$dbName' finding attachment '$attachmentName' of document '$doc'")
+
+    require(doc != null, "doc undefined")
+    require(doc.rev.rev != null, "doc revision must be specified")
+
+    val downloadStream = gridFSBucket.openDownloadStream(BsonString(s"${doc.id.id}/$attachmentName"))
+
+    def readStream(file: GridFSFile) = {
+      val source = MongoDBAsyncStreamSource(downloadStream)
+      source
+        .runWith(sink)
+        .map { result =>
+          transid
+            .finished(
+              this,
+              start,
+              s"[ATT_GET] '$collName' completed: found attachment '$attachmentName' of document '$doc'")
+          result
+        }
+    }
+
+    def getGridFSFile = {
+      downloadStream
+        .gridFSFile()
+        .head()
+        .transform(
+          identity, {
+            case ex: MongoGridFSException if ex.getMessage.contains("File not found") =>
+              transid.finished(
+                this,
+                start,
+                s"[ATT_GET] '$collName', retrieving attachment '$attachmentName' of document '$doc'; not found.")
+              NoDocumentException("Not found on 'readAttachment'.")
+            case ex: MongoGridFSException =>
+              transid.failed(
+                this,
+                start,
+                s"[ATT_GET] '$collName' failed to get attachment '$attachmentName' of document '$doc'; error code: '${ex.getCode}'",
+                ErrorLevel)
+              throw new Exception("Unexpected mongodb server error: " + ex.getMessage)
+            case t => t
+          })
+    }
+
+    val f = for {
+      file <- getGridFSFile
+      result <- readStream(file)
+    } yield result
+
+    reportFailure(
+      f,
+      failure =>
+        transid.failed(
+          this,
+          start,
+          s"[ATT_GET] '$dbName' internal error, name: '$attachmentName', doc: '$doc', failure: '${failure.getMessage}'",
+          ErrorLevel))
+
+  }
+
+  override protected[core] def deleteAttachments[T](doc: DocInfo)(implicit transid: TransactionId): Future[Boolean] =
+    attachmentStore
+      .map(as => as.deleteAttachments(doc.id))
+      .getOrElse(Future.successful(true)) // For MongoDB it is expected that the entire document is deleted.
+
+  override def shutdown(): Unit = {
+    // MongoClient maintains the connection pool internally, we don't need to manage it
+    attachmentStore.foreach(_.shutdown())
+  }
+
+  private def reportFailure[T, U](f: Future[T], onFailure: Throwable => U): Future[T] = {
+    f.failed.foreach {
+      case _: ArtifactStoreException => // These failures are intentional and shouldn't trigger the catcher.
+      case x                         => onFailure(x)
+    }
+    f
+  }
+
+  // calculate the revision manually, to be compatible with couchdb's _rev field
+  private def revisionCalculate(doc: JsObject): (String, String) = {
+    val md: MessageDigest = MessageDigest.getInstance("MD5")
```

Comment:
minor nit: I think we can use store util instead?
https://github.com/apache/openwhisk/blob/master/common/scala/src/main/scala/org/apache/openwhisk/core/database/StoreUtils.scala#L101

Should the underlying algorithm be MD5?

Reply:
great, I will update